internal/curator/durable/state/fb/extra.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2017 Western Digital Corporation or its affiliates. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package fb
+
+import (
+	flatbuffers "github.com/google/flatbuffers/go"
+
+	"github.com/westerndigitalcorporation/blb/internal/core"
+)
+
+// Helpers for meta in BlobF:
+
+const (
+	metaStorage = iota * 8
+	metaHint
+	metaRepl
+)
+
+func (b *BlobF) Storage() core.StorageClass {
+	return core.StorageClass((b.PackedMeta() >> metaStorage) & 0xff)
+}
+
+func (b *BlobF) Hint() core.StorageHint {
+	return core.StorageHint((b.PackedMeta() >> metaHint) & 0xff)
+}
+
+func (b *BlobF) Repl() int {
+	return int((b.PackedMeta() >> metaRepl) & 0xff)
+}
+
+func PackMeta(storage core.StorageClass, hint core.StorageHint, repl int) uint32 {
+	return (uint32(storage)&0xff)<<metaStorage |
+		(uint32(hint)&0xff)<<metaHint |
+		(uint32(repl)&0xff)<<metaRepl
+}
+
+// Expand/truncate unix nanos/seconds so we can work with nanos in the rest of
+// the codebase:
+
+func (b *BlobF) Deleted() int64 { return int64(b.Deletedsec()) * 1e9 }
+func (b *BlobF) Mtime() int64   { return int64(b.Mtimesec()) * 1e9 }
+func (b *BlobF) Atime() int64   { return int64(b.Atimesec()) * 1e9 }
+func (b *BlobF) Expires() int64 { return int64(b.Expiressec()) * 1e9 }
+
+func (b *BlobF) MutateDeleted(n int64) bool { return b.MutateDeletedsec(uint32(n / 1e9)) }
+func (b *BlobF) MutateMtime(n int64) bool   { return b.MutateMtimesec(uint32(n / 1e9)) }
+func (b *BlobF) MutateAtime(n int64) bool   { return b.MutateAtimesec(uint32(n / 1e9)) }
+func (b *BlobF) MutateExpires(n int64) bool { return b.MutateExpiressec(uint32(n / 1e9)) }
+
+func BlobFAddDeleted(b *flatbuffers.Builder, n int64) { BlobFAddDeletedsec(b, uint32(n/1e9)) }
+func BlobFAddMtime(b *flatbuffers.Builder, n int64)   { BlobFAddMtimesec(b, uint32(n/1e9)) }
+func BlobFAddAtime(b *flatbuffers.Builder, n int64)   { BlobFAddAtimesec(b, uint32(n/1e9)) }
+func BlobFAddExpires(b *flatbuffers.Builder, n int64) { BlobFAddExpiressec(b, uint32(n/1e9)) }
+
+// Helpers for TractIDF:
+
+func (t *TractIDF) TractID() core.TractID {
+	bid := uint64(t.B5())<<48 | uint64(t.B4())<<32 | uint64(t.B3())<<16 | uint64(t.B2())
```

Comment:
question how you approach bit shifting code? What do you do to make sure this is right aside from testing? Do you draw pictures? Bitshifting code I've had to write in OS course stuff has been tiny.

Reply:
Tests are a good idea, and feel free to suggest adding tests where you think there's value or things aren't trivial.

Personally, it depends how tricky it is: for these particular functions, they're all on nice byte boundaries so it feels easy enough. Some trickier ones, I'll do a little diagram with ascii art. I think Go is good for bitwise operations in general since it strictly enforces casts between integer widths. It's annoying but I find it does prevent certain kinds of bugs.

Reply:
Yeah. I think tests are a good idea here. Feel free to add more tests. The code looks right and it's simple, but maybe some simple tests could sanity check that IDs are extracted properly.

Reply:
Yup, I think the best way to do these is test that BuildX and XF.ToStruct are inverses (when various combinations of fields are set). That should exercise all this code.